docs/rfc/rfc-017-abci++-vote-extension-propag.md

```diff
@@ -0,0 +1,438 @@
+  reconstruct the votes with their corresponding extensions
+
+The changes to the blocksync reactor are more substantial:
+
+- the `BlockResponse` message is extended to include the extended commit
+- structure `bpRequester` is likewise extended to hold the received extended commits coming in
+  `BlockResponse` messages
+- method `PeekTwoBlocks` is modified to also return the extended commit corresponding to the first block
+- when successfully verifying a received block, the reactor saves its corresponding extended commit in
+  the block store
+
+The two main drawbacks of this base implementation are:
+
+- the increased size taken by the block store, in particular with big extensions
+- the increased bandwith taken by the new format of `BlockResponse`
+
+#### Possible Optimization: Pruning the Extended Commit History
+
+If we cannot switch from the consensus reactor back to the blocksync reactor, and we do not have
+the catch-up mechanism in the consensus reactor explained above, we cannot prune the extended commit
+backlog in the block store without sacrificing the implementation's correctness. The asynchronous
+nature of our distributed system model allows a process to fall behing an arbitrary number of
+heights, and thus all extended commits need to be kept *just in case* a node that late had
+previously switched to the consensus reactor.
+
+However, there is a possibility to optimize the base implementation. We could prune from the block
+store all extended commits that are more than *d* heights in the past. Then, we need to handle two
+new situations, roughly equivalent to cases (g.1) and (g.2) described above.
+
+- (g.1) A node starts from scratch or recovers after a crash, and (for simplicity of explanation)
+  it does not run statesync.
+  In this case, the blocksync reactor needs to be improved to make sure it cannot switch to consensus
+  until it has received a valid extended commit for a height in its future. This extended commit
+  will allow the node to switch from the blocksync reactor to the consensus reactor and immediately
+  act as a proposer if required.
+- (g.2) A node already running the consensus reactor falls behind beyond *d* heights. In principle,
+  the node will be stuck as no other node can provide the vote extensions it needs to make progress.
+  However we can have the node crash and recover as a workaround. This effectively converts this case
+  into the previous one.
+
+### Formalization Work
+
+A formalization work to show or prove the correctness of the different use cases and solutions
+presented here (and any other that may be found) needs to be carried out.
+A question that needs a precise answer is how many extended commits (one?, two?) a node needs
+to persist when implementing solution 3 without Tendermint's current limitations.
+Another important invariant we need to prove formally is that the set of vote extensions
+required to make progress will always be held somewhere in the network.
+
```

Comment:
A general point here. Block sync is a p2p protocol, where a node interacts with peers (nodes it is connected directly to). So, the backlog a node needs to store depends only on the current height of all its peers, not the entire network. In words, it is a local computation, that depends on the state only of the peers of a node.

Moreover, I can't really formalize that, but I consider that we don't need either to keep the "lowest height" among peers, which is somehow the assumption in this RFC. More precisely, we can have a peer joining or rejoining the network, this peer is likely to be very late, but for most of the block we need to provide for this peer, the vote extensions are not needed, from the simple point that this node is the most outdated among its peers, the only nodes it interacts with.

Reply:
I think I agree with this. The crux of Solution 3 is that outdated nodes don't need extensions to catch up. However, nodes catching up need to have the guarantee that, if they advance a lot of heights by blocksync (so, no extensions, just applying blocks by calling `FinalizeBlock`), they have the _guarantee_ that there is an extended commit in their "future" that will allow them to "switch to consensus" (stated in Tendermint reactor terms).
Let me know if you were expecting me to add some text here regarding this.